### ISSUE TITLE
pods from crashed slaves linger indefinitely in apiserver's registry

### ISSUE THREAD

**Issue by jdef** [mesosphere/kubernetes-mesos#158]
kubernetes: v0.11

To reproduce:
* create a replicated pod across multiple slaves, wait for pod instances to come up everywhere
* terminate some slave nodes where the pod is running
* wait a couple of minutes for the controller-manager's node controller to realize that the nodes are down, and attempt to delete the pods
  * _observe that deleting the pods fails_
  * pods are still visible via the apiserver REST interface
  * pods are still "bound" upon manual inspection of etcd `registry/nodes/{node}/boundpods/`

Note:
* unclear why the pod delete fails: v0.11 does not log the error (but it looks like v0.12 will)
* workaround: it is possible to manually delete the pods via the apiserver's REST interface (tested via `curl`)

Some logs:
```
[scheduler.log]
I0303 18:13:41.408071   28156 messenger.go:166] Sending message mesos.internal.LaunchTasksMessage to master@10.223.113.227:5050
I0303 18:13:41.409019   28156 messenger.go:343] Receiving message mesos.internal.StatusUpdateMessage from master@10.223.113.227:5050
I0303 18:13:41.409082   28156 scheduler.go:562] Received status update from  master@10.223.113.227:5050  status source: @0.0.0.0:0
I0303 18:13:41.409098   28156 scheduler.go:210] Received status update &TaskStatus{TaskId:&TaskID{Value:*41952bf9-c1a7-11e4-be18-42010adf71e3,XXX_unrecognized:[],},State:*TASK_LOST,Message:*Slave 10.150.52.19 disconnected,Source:*SOURCE_MASTER,Reason:*REASON_SLAVE_DISCONNECTED,Data:nil,SlaveId:&SlaveID{Value:*20150221-134408-3815890698-5050-18127-S2,XXX_unrecognized:[],},ExecutorId:nil,Timestamp:*1.425406421403547e+09,Healthy:nil,XXX_unrecognized:[],}
W0303 18:13:41.409179   28156 registry.go:243] task lost: &TaskStatus{TaskId:&TaskID{Value:*41952bf9-c1a7-11e4-be18-42010adf71e3,XXX_unrecognized:[],},State:*TASK_LOST,Message:*Slave 10.150.52.19 disconnected,Source:*SOURCE_MASTER,Reason:*REASON_SLAVE_DISCONNECTED,Data:nil,SlaveId:&SlaveID{Value:*20150221-134408-3815890698-5050-18127-S2,XXX_unrecognized:[],},ExecutorId:nil,Timestamp:*1.425406421403547e+09,Healthy:nil,XXX_unrecognized:[],}
I0303 18:13:41.409233   28156 scheduler.go:581] Sending status update ACK to  master@10.223.113.227:5050
I0303 18:13:41.409265   28156 messenger.go:166] Sending message mesos.internal.StatusUpdateAcknowledgementMessage to master@10.223.113.227:5050

[apiserver.log]
E0303 18:15:04.815207    6121 server.go:328] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
E0303 18:15:14.666813    6121 pod_cache.go:127] Unexpected error verifying node existence: minion "10.22.211.18" not found
E0303 18:15:19.708782    6121 pod_cache.go:127] Unexpected error verifying node existence: minion "10.22.211.18" not found
E0303 18:15:19.815483    6121 server.go:328] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
E0303 18:15:24.763088    6121 pod_cache.go:127] Unexpected error verifying node existence: minion "10.22.211.18" not found

[controller.log]
E0303 18:14:56.039046    6165 endpoints_controller.go:81] Failed to find a host IP for pod default/redis-slave-controller-5c2ib
E0303 18:15:06.095505    6165 endpoints_controller.go:81] Failed to find a host IP for pod default/redis-slave-controller-5c2ib
I0303 18:15:10.547499    6165 nodecontroller.go:184] Delete node from registry: 10.22.211.18
E0303 18:15:10.594827    6165 nodecontroller.go:343] Error deleting pod redis-slave-controller-5c2ib
E0303 18:15:16.157511    6165 endpoints_controller.go:81] Failed to find a host IP for pod default/redis-slave-controller-5c2ib
I0303 18:15:22.308987    6165 nodecontroller.go:184] Delete node from registry: 10.150.52.19
E0303 18:15:22.363998    6165 nodecontroller.go:343] Error deleting pod frontend-controller-zzsm4
E0303 18:15:26.220227    6165 endpoints_controller.go:81] Failed to find a host IP for pod default/redis-slave-controller-5c2ib
```

**Comment by jdef** [comment]
Example of lingering pod, via `curl http://$servicehost:8888/api/v1beta2/pods/redis-slave-controller-5c2ib`:
```json
{
  "kind": "Pod",
  "id": "redis-slave-controller-5c2ib",
  "uid": "57b68280-c17d-11e4-bde2-42010adf71e3",
  "creationTimestamp": "2015-03-03T08:14:41Z",
  "selfLink": "/api/v1beta2/pods/redis-slave-controller-5c2ib?namespace=default",
  "resourceVersion": 839148,
  "apiVersion": "v1beta2",
  "namespace": "default",
  "generateName": "redis-slave-controller-",
  "labels": {
    "name": "redisslave"
  },
  "desiredState": {
    "manifest": {
      "version": "v1beta2",
      "id": "",
      "volumes": null,
      "containers": [
        {
          "name": "slave",
          "image": "jdef/redis-slave",
          "ports": [
            {
              "hostPort": 31020,
              "containerPort": 6379,
              "protocol": "TCP"
            }
          ],
          "resources": {},
          "terminationMessagePath": "/dev/termination-log",
          "imagePullPolicy": "PullIfNotPresent",
          "capabilities": {}
        }
      ],
      "restartPolicy": {
        "always": {}
      },
      "dnsPolicy": "ClusterFirst"
    }
  },
  "currentState": {
    "manifest": {
      "version": "",
      "id": "",
      "volumes": null,
      "containers": null,
      "restartPolicy": {}
    },
    "status": "Unknown",
    "host": "10.22.211.18"
  }
}
```

Etcd registry state:
```json
{"action":"get",
"key":"/registry/pods/default/redis-slave-controller-5c2ib",
"value":
    "{\"kind\":\"Pod\",
    \"id\":\"redis-slave-controller-5c2ib\",
    \"uid\":\"57b68280-c17d-11e4-bde2-42010adf71e3\",
    \"creationTimestamp\":\"2015-03-03T08:14:41Z\",
    \"resourceVersion\":839051,
    \"apiVersion\":\"v1beta1\",
    \"namespace\":\"default\",
    \"generateName\":\"redis-slave-controller-\",
    \"labels\":{\"name\":\"redisslave\"},
    \"desiredState\":{
        \"manifest\":{
            \"version\":\"v1beta2\",
            \"id\":\"\",\"volumes\":null,
            \"containers\":[{
                \"name\":\"slave\",
                \"image\":\"jdef/redis-slave\",
                \"ports\":[{
                    \"hostPort\":31020,\"containerPort\":6379,\"protocol\":\"TCP\"
                }],
                \"resources\":{},
                \"terminationMessagePath\":\"/dev/termination-log\",
                \"imagePullPolicy\":\"PullIfNotPresent\",
                \"capabilities\":{}
            }],
            \"restartPolicy\":{\"always\":{}},
            \"dnsPolicy\":\"ClusterFirst\"
        }
    },
    \"currentState\":{
        \"manifest\":{
            \"version\":\"\",
            \"id\":\"\",
            \"volumes\":null,
            \"containers\":null,
            \"restartPolicy\":{}
        },
        \"status\":\"Waiting\",
        \"host\":\"10.22.211.18\"
    }
}",
"index":911184}
```

**Comment by jdef** [comment]
Reproduced the problem and attempted the work-around immediately, got a 404 error - suspect something related to the pod cache. However, the pod was actually deleted!

```shell
$ curl -v http://$servicehost:8888/api/v1beta1/pods/frontend-controller-yhrwx -XDELETE
* About to connect() to 10.223.113.227 port 8888 (#0)
*   Trying 10.223.113.227...
* connected
* Connected to 10.223.113.227 (10.223.113.227) port 8888 (#0)
> DELETE /api/v1beta1/pods/frontend-controller-yhrwx HTTP/1.1
> User-Agent: curl/7.26.0
> Host: 10.223.113.227:8888
> Accept: */*
>
* additional stuff not fine transfer.c:1037: 0 0
* HTTP 1.1 or later with persistent connection, pipelining supported
< HTTP/1.1 404 Not Found
< Content-Type: application/json
< Date: Tue, 03 Mar 2015 19:53:19 GMT
< Content-Length: 279
<
{
  "kind": "Status",
  "creationTimestamp": null,
  "apiVersion": "v1beta1",
  "status": "Failure",
  "message": "pods \"frontend-controller-yhrwx\" not found",
  "reason": "NotFound",
  "details": {
    "id": "frontend-controller-yhrwx",
    "kind": "pods"
  },
  "code": 404
* Connection #0 to host 10.223.113.227 left intact
}* Closing connection #0
```

**Comment by jdef** [comment]
Patched the kubernetes code (#159) and now I see this in the controller.log:
```
E0303 19:59:00.441721   14392 nodecontroller.go:343] Error deleting pod frontend-controller-zy30y: request construction error: 'an empty namespace may not be set when a resource name is provided'
E0303 19:59:00.441781   14392 nodecontroller.go:343] Error deleting pod redis-slave-controller-haca3: request construction error: 'an empty namespace may not be set when a resource name is provided'
```

**Comment by jdef** [comment]
FWIW - this is fixed in upstream/master, but not in v0.12

**Comment by jdef** [comment]
resolved by #161
